File: JavaSimpleBackend/src/main/java/com/atesti/controller/RadniNaloziController.java
package com.atesti.controller;

import com.atesti.dto.AssignWorkerRequest;
import com.atesti.dto.RadniNalogRequest;
import com.atesti.dto.UskoroIsticeResponse;
import com.atesti.entity.RadniNalog;
import com.atesti.service.RadniNalogPdfService;
import com.atesti.service.RadniNaloziService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/radni-nalozi")
@RequiredArgsConstructor
public class RadniNaloziController {

    private final RadniNaloziService radniNaloziService;
    private final RadniNalogPdfService radniNalogPdfService;

    @GetMapping("/rn001.pdf")
    public ResponseEntity<byte[]> downloadRn001Pdf() {
        Map<String, Object> result = radniNalogPdfService.generateRadniNalogPdf(1L);
        return buildPdfResponse(result, "RN001.pdf");
    }

    @GetMapping("/nextBrojNaloga")
    public ResponseEntity<Map<String, String>> getNextBrojNaloga() {
        String next = radniNaloziService.getNextBrojNaloga();
        return ResponseEntity.ok(Map.of("brojNaloga", next));
    }

    @GetMapping("/uskoroIstice")
    public ResponseEntity<List<UskoroIsticeResponse>> getUskoroIstice(
            @RequestParam(required = false) Integer days) {
        int daysValue = (days != null) ? days : 1000;
        return ResponseEntity.ok(radniNaloziService.getUskoroIstice(daysValue));
    }

    @GetMapping
    public ResponseEntity<List<RadniNalog>> getAll() {
        return ResponseEntity.ok(radniNaloziService.getAllRadniNalozi());
    }

    @GetMapping("/{id}")
    public ResponseEntity<RadniNalog> getById(@PathVariable Long id) {
        return ResponseEntity.ok(radniNaloziService.getRadniNalogById(id));
    }

    @GetMapping("/{id}/pdf")
    public ResponseEntity<byte[]> downloadPdf(@PathVariable Long id) {
        Map<String, Object> result = radniNalogPdfService.generateRadniNalogPdf(id);
        String fileName = (String) result.get("fileName");
        return buildPdfResponse(result, fileName);
    }

    @PostMapping
    public ResponseEntity<RadniNalog> create(@RequestBody RadniNalogRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(radniNaloziService.createRadniNalog(request));
    }

    @PutMapping("/{id}")
    public ResponseEntity<RadniNalog> update(@PathVariable Long id, @RequestBody RadniNalogRequest request) {
        return ResponseEntity.ok(radniNaloziService.updateRadniNalog(id, request));
    }

    @PutMapping("/{id}/assign")
    public ResponseEntity<RadniNalog> assignWorker(@PathVariable Long id, @RequestBody AssignWorkerRequest request) {
        return ResponseEntity.ok(radniNaloziService.assignWorker(id, request.getAssignedUserId()));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<RadniNalog> delete(@PathVariable Long id) {
        return ResponseEntity.ok(radniNaloziService.deleteRadniNalog(id));
    }

    private ResponseEntity<byte[]> buildPdfResponse(Map<String, Object> result, String fileName) {
        byte[] buffer = (byte[]) result.get("buffer");
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_PDF);
        headers.setContentDispositionFormData("attachment", fileName);
        return new ResponseEntity<>(buffer, headers, HttpStatus.OK);
    }
}

File: backend/controllers/radniNaloziController.js
const radniNaloziService = require('../services/radniNaloziService');
const radniNalogPdfService = require('../services/radniNalogPdfService');

class RadniNaloziController {
    async getNextBrojNaloga(req, res) {
      try {
        const nextBroj = await radniNaloziService.getNextBrojNaloga();
        res.json({ brojNaloga: nextBroj });
      } catch (error) {
        res.status(500).json({ message: error.message });
      }
    }
  async getUskoroIstice(req, res) {
    try {
      const parsedDays = parseInt(req.query.days, 10);
      const days = Number.isNaN(parsedDays) ? 1000 : parsedDays;
      const uskoroIstice = await radniNaloziService.getUskoroIstice(days);
      const response = uskoroIstice.map((item) => ({
        ...item,
        isActive: item.isActive !== undefined ? item.isActive : true
      }));
      res.json(response);
    } catch (error) {
      res.status(500).json({ message: error.message });
    }
  }

  async getAll(req, res) {
    try {
      const nalozi = await radniNaloziService.getAllRadniNalozi();
      res.json(nalozi);
    } catch (error) {
      res.status(500).json({ message: error.message });
    }
  }

  async getById(req, res) {
    try {
      const nalog = await radniNaloziService.getRadniNalogById(parseInt(req.params.id));
      res.json(nalog);
    } catch (error) {
      res.status(404).json({ message: error.message });
    }
  }

  async getByNarucitelj(req, res) {
    try {
      const nalozi = await radniNaloziService.getRadniNaloziByNarucitelj(parseInt(req.params.naruciteljiId));
      res.json(nalozi);
    } catch (error) {
      res.status(500).json({ message: error.message });
    }
  }

  async create(req, res) {
    try {
      console.log('Creating Radni Nalog with data:', req.body);
      const nalog = await radniNaloziService.createRadniNalog(req.body);
      res.status(201).json(nalog);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  async update(req, res) {
    try {
      const nalog = await radniNaloziService.updateRadniNalog(parseInt(req.params.id), req.body);
      res.json(nalog);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  async assignWorker(req, res) {
    try {
      const { assignedUserId } = req.body;
      const nalog = await radniNaloziService.assignWorker(parseInt(req.params.id), assignedUserId);
      res.json(nalog);
    } catch (error) {
      res.status(400).json({ message: error.message });
    }
  }

  async delete(req, res) {
    try {
      const nalog = await radniNaloziService.deleteRadniNalog(parseInt(req.params.id));
      res.json(nalog);
    } catch (error) {
      res.status(404).json({ message: error.message });
    }
  }

  async downloadPdf(req, res) {
    try {
      const result = await radniNalogPdfService.generateRadniNalogPdf(parseInt(req.params.id));
      res.setHeader('Content-Type', result.contentType);
      res.setHeader('Content-Disposition', `attachment; filename="${result.fileName}"`);
      res.send(result.buffer);
    } catch (error) {
      res.status(404).json({ message: error.message });
    }
  }

  async downloadRn001Pdf(req, res) {
    try {
      const result = await radniNalogPdfService.generateRadniNalogPdf(1);
      res.setHeader('Content-Type', result.contentType);
      res.setHeader('Content-Disposition', 'attachment; filename="RN001.pdf"');
      res.send(result.buffer);
    } catch (error) {
      res.status(404).json({ message: error.message });
    }
  }
}

module.exports = new RadniNaloziController();
